Automattic/wp-calypso
# Fetch translation chunks plugin

#### Changes proposed in this Pull Request

This PR introduces the initial implementation of using chunked translations. When this feature is enabled, Calypso will switch from fetching a single file, that contains the complete set of strings, to fetching separate translation chunks bundled with each Calypso build.

In order to achieve that, it adds a few additional steps to the build process, starting with generating chunks map. Chunks map is a JSON file, that contains the paths to all JS modules keyed by the chunk they were imported to, generated by a webpack plugin.

Next step is to download `calypso-strings.pot` and translation JSON files for all languages from `https://widgets.wp.com/languages/calypso`. Previously generated chunks map is used together with `calypso-strings.pot` to get the set of strings per chunk based on the reference comments in the POT file, which is then used with the actual translations to generate individual translation chunks for each language. 

Another webpack plugin will inject additional code to [The Manifest](https://webpack.js.org/concepts/manifest/) that will expose globally a callback called whenever webpack attempts to require a JS chunk. `__requireChunkCallback__` object is attached to the window object and used to fetch translation chunks in the runtime. 

On boot, the language manifest for user's language will be downloaded. It contains the locale data and a set of all chunk ids that have translation chunks and is used to filter translation chunk request that don't exist. Then, an event handler is attached to the exposed require chunk callback to fetch translation chunks as whenever a JS chunk is being fetched in the runtime.

This is an experimental feature and therefore it won't be enabled by default. Production Docker build will also build translation chunks, but client will be still using the standard translation files, unless `use-translation-chunks` is enabled via URL flag (e.g. adding `?flags=use-translation-chunks` to the URL).

#### Testing instructions

- Boot Calypso with `ENABLE_FEATURES=use-translation-chunks npm start`
- Open http://calypso.localhost:3000/

See p4TIVU-9nl-p2
